Implemented configurable listening address

elluisian opened the pull request:
When applied, this commit tries to implement the feature requests reported on the issues #43 and #227. Basically, with this, it is now possible to make the server listen only on certain addresses. This increases security and allows the typical "SSH usage", that is, SSH + publickey auth + local port forwarding + VNC listening on localhost only.

To achieve this:

- on droidvnc-ng.c, vncServerStart was modified in order to provide a further "jstring listenIf" parameter. It uses rfbScreenInfo*'s listenInterface property.

- A TableRow was added in order to allow users to input the desired address to use ("Listening Address").

- On droidvnc-ng.c, the method vncServerGetListenInterface was introduced in order to track if the server is currenctly set to listen to 0.0.0.0 or not (this is used to properly show what addresses are available to use on the UI).

Please note that, if an invalid address/host is set, by default, it is assumed the address 0.0.0.0 is requested.

bk138 commented:
I think we should aim for "interface names", not "IP addresses" in the UI. Also, what happens if the interface the server is using the IP of goes down?

elluisian commented:
@bk138 Hi, sorry for the late response.
Actually, I think I've used "listening address" as label... not IP address, also, technically, one specifies a single address of an interface to listen to (except for 0.0.0.0 which is a special address which stands for "all of them").

As for what happens when the interface is down... I'm sorry, I didn't check for that when I made the fix, I've checked while commenting this though.

I've made the server listening to my wifi IP Address and, if I disable wifi while the server is running, it simply stops (I guess this is the default behaviour you programmed?), if I do the same thing but when listening on localhost, nothing happens, the server continues to listen, same thing if using 0.0.0.0.

bk138 commented:
What I mean is that we should not bother the user to enter IP addresses in the UI but present choices like "Wifi interface", LAN, WAN etc and handle the device->address resolution under the hood.

elluisian commented:
How about a spinner with the current up interfaces plus, the possibility to define a custom IP address?
Something like:

Wifi
Cellular
Loopback
...
Custom IP

It seems XVnc has a -interface option in order to customize the ip address.

bk138 commented:
Yes, but why would the user want to set a custom IP?

elluisian commented:
Ehr... eh eh, that's actually a good question... because... why not? :P
I know, stupid answer... the only use-case I can think of is if one wants to use a loopback address which is DIFFERENT FROM 127.0.0.1 (for example, 127.0.0.124).
I understand this is a far-fetched use-case... but still, considering it is relatively easy to make, why not?
All that is needed is to simply make the spinner and an edittext, similar to the one in my commit, only, this time, it is enabled only if "custom ip" is chosen.

bk138 commented:
Let's keep it as simple as needed: https://en.wikipedia.org/wiki/KISS_principle :-)

elluisian commented:
Ok, done, now it should be better, let me know what you think.

bk138 commented on app/src/main/cpp/droidvnc-ng.c:
```diff
@@ -315,6 +315,17 @@ JNIEXPORT jboolean JNICALL Java_net_christianbeier_droidvnc_1ng_MainService_vncS
 
+    in_addr_t address = 0; // Default is 0.0.0.0
+    if (listenIf != NULL) {
+        const char *listenIfNative = (*env)->GetStringUTFChars(env, listenIf, NULL);
```
better yet, name it cListenInterface. If reads too much like if.

bk138 commented on app/src/main/java/net/christianbeier/droidvnc_ng/Defaults.kt:
```diff
@@ -40,6 +40,10 @@ class Defaults {
     }
 
+    @EncodeDefault
+    var listenInterface = "0.0.0.0"
```
Comment whether this be an IP address or an interface name.

elluisian replied:
Since we're aiming for interface names, the idea is that no interface should be named "0.0.0.0", basically, this should represent a string that simply states "All interfaces" (I know, quite confusing).
Do you have a better idea for this? I don't like it too much myself, except that... I don't have a better one.

bk138 commented on app/src/main/java/net/christianbeier/droidvnc_ng/ListenIfAdapter.java:
```diff
@@ -0,0 +1,263 @@
+            this.nic = nic;
+
+            if (nic == null) {
+                this.techName = "0.0.0.0";
```
Just use one name please.

bk138 commented on app/src/main/java/net/christianbeier/droidvnc_ng/ListenIfAdapter.java:
```diff
@@ -0,0 +1,263 @@
+     * @param ifName name (technical) of the interface
+     * @return if known, a friendly name for the interface, otherwise, null
+     */
+    private String getFriendlyStringForInterface(String ifName) {
```
This manual mapping gets us (or rather me) into a rabbithole of users complaining later on that their exotic device isn't listed. Please just use https://developer.android.com/reference/java/net/NetworkInterface#getDisplayName()

elluisian replied:
I kept a copy of this just in case, I'll consider it solved for now.

bk138 commented on app/src/main/java/net/christianbeier/droidvnc_ng/MainService.java:
```diff
@@ -753,12 +766,88 @@ static boolean isServerActive() {
+     * Get all the available network interfaces, in the up status, which have at least one IPv4 address
+     * @return An ArrayList&lt;NetworkInterfaces&gt; containing the above-mentioned interfaces
+     */
+    static ArrayList<NetworkInterface> getAvailableNICs() {
```
Please move into Utils.kt

bk138 commented on app/src/main/java/net/christianbeier/droidvnc_ng/MainService.java:
```diff
@@ -753,12 +766,88 @@ static boolean isServerActive() {
+     * @param ni the NetworkInterface to check
+     * @return an ArrayList&lt;String&gt; which contains all the collected ipv4
+     */
+    static ArrayList<String> getIPv4ForInterface(NetworkInterface ni) {
```
Same here, better in Utils.

bk138 commented on app/src/main/java/net/christianbeier/droidvnc_ng/MainService.java:
```diff
@@ -753,12 +766,88 @@ static boolean isServerActive() {
+     * Get all available IPv4 addresses for reaching the current running server.
+     * @return A list of strings, each containing one IPv4 address.
+     */
+    static ArrayList<String> getReachableIPv4s() {
```
Can also move to Utils I think

elluisian replied:
I don't think this is a good idea since there's a call to getListenInterface(), which in turns, uses SharedPreferences.
Moreover, there's some specific Chrome OS code.

bk138 commented on app/src/main/java/net/christianbeier/droidvnc_ng/MainService.java:
```diff
@@ -770,29 +859,45 @@ static ArrayList<String> getIPv4s() {
+            SharedPreferences prefs = PreferenceManager.getDefaultSharedPreferences(instance);
+            return prefs.getString(PREFS_KEY_SERVER_LAST_LISTEN_INTERFACE, new Defaults(instance).getListenInterface());
+        } catch (Exception e) {
+            return null;
```
Rather "0.0.0.0"? Unsure...

elluisian replied:
What do you suggest?

elluisian replied:
Changed default interface to loopback, this way, it is safer.

bk138 commented on app/src/main/java/net/christianbeier/droidvnc_ng/OnBootReceiver.java:
```diff
@@ -54,6 +54,7 @@ public void onReceive(Context context, Intent arg1)
 
             Intent intent = new Intent(context, MainService.class);
             intent.setAction(MainService.ACTION_START);
+            intent.putExtra(MainService.EXTRA_LISTEN_INTERFACE, prefs.getString(Constants.PREFS_KEY_SETTINGS_LISTEN_INTERFACE, defaults.getListenInterface()));
```
We cannot be sure the interface is up at that time. How to handle?

elluisian replied:
How about something like this on the Java-side of things?
https://stackoverflow.com/questions/8919083/checking-host-reachability-availability-in-android.
It may solve the problem directly from Java, without touching the .c file at all.

elluisian replied:
I believe that it is important to first solve the problem of intercepting network change states, after that, this should be easier.

elluisian replied:
I've made an implementation, I haven't tested it yet though, for some reason, I'm not able to run the server at boot.
Do I need to do something in particular? I'm using a Xiaomi for testing.

elluisian replied:
Now OnBootReceiver sends a notifications when the interface is not up, avoiding, this way, to start the service.

bk138 commented on libvncserver:
This should not be part of the PR.

elluisian replied:
Sorry, I know, I'm still a bit unexperienced... I was trying to make my local copy up-to-date with your newest changes... do you have any suggestion on how to avoid that in the future?

bk138 replied:
Yes ;-) A git diff to the master branch and careful looking :-)

elluisian replied:
You mean, cherry-pick single commits?

bk138 reviewed (changes requested):
Some comments. Before delving into details, the two grand topics are:

- what's the strategy when a listen interface goes down while the server is running?
- what's the strategy when the server is started wit a listen interface that's not up?

If solutions for these are too hard to be found, the details are kinda moot...

elluisian commented on app/src/main/java/net/christianbeier/droidvnc_ng/ListenIfAdapter.java:
```diff
@@ -0,0 +1,263 @@
+/*
+ * DroidVNC-NG ListenIfAdapter.
+ *
+ * Author: Christian Beier <info@christianbeier.net
```
Sorry, copy and paste does that to you :P

elluisian commented on app/src/main/cpp/droidvnc-ng.c:
```diff
@@ -315,6 +315,17 @@ JNIEXPORT jboolean JNICALL Java_net_christianbeier_droidvnc_1ng_MainService_vncS
+    }
+
+    // With the listenInterface property one can define where the server will be available
+    theScreen->listenInterface = address;
```
I have to be honest, on this side I had no idea, but, looking at the source code for libvnc, it seems that internally, rfbServerInit uses rfbListenOnTCPPort for "auto-probing" at some point... we could use that to verify if the interface is up, I guess.

elluisian replied:
Hi, I tried working on listening for Network events, but I'm having some problems.
I found a way to intercept Network State changes, but it still doesn't work as expected (for example, when I turn on/off the VPN it doesn't work), moreover, I don't know how to integrate it with the service, for now, only WLAN works for me...

The NetworkInterfaceTester is used, at the moment, only to update the spinner, but with some more work, we should be capable of using it to properly intercept network changes even in the service.

I think I need help though, I don't know how to integrate this "tester" with the Service (I still haven't learnt how to deal with those in Android).
Of course, even only suggestions would do, I'm not expecting others to do the work for me.

As for the .c file, if the address is not valid, an error is reported (instead of using 0.0.0.0).
I'm not sure how to deal with check if the interface is up in the c-side of things... the actual problem, is detecting the change in the used network interface.

bk138 replied:
Maybe you don't have to handle the situation - if the server just keeps running and serves requests again after the interface is up again, that'd be OK. It should just not get into an unrecoverable state. I have not tested this at all, so just asking what you see :-)

elluisian replied:
Oh, do not worry... actually, I was thinking of another approach on the c-side of things, except, well, I'm not well to informed with libvnc.
Is there a way to detect connections problems with it? Something like stopping rfbRunEventLoop and set some flag?

elluisian replied:
As far as I can tell, the function vncIsActive always reports true, when listening to WLAN, even though Wi-Fi is removed.
Internally, vncIsActive uses rfbIsActive from libvnc which, it seems it does not take into account the listenInterface "attribute" of RfbScreen.

How about some kind of polling mechanism in order to see if the socket (listenSock in our case) is still "usable"?

elluisian replied:
I can think of two possible ways to solve this:

1. Java-only: detect Network state change, if the server is listening to the interface that was removed, force-stop the server with vncServerStop;
2. C+Java: make it possible to automatically stop the server from C, by checking the connection state and stopping the rfb thread accordingly.
I guess network state detection may be used to update the UI only, since everything is done from C.

elluisian replied:
I was able to implement a Java-only solution, it probably needs more polish though.
Also, for some reason, it didn't work with my VPN (mullvad), the callbacks were not called at all.
I tested with Wifi though, it should work with cellular data also.

elluisian replied:
Problem with VPN was solved, it should work with any interface now.
I'm waiting for review.

elluisian commented on app/src/main/java/net/christianbeier/droidvnc_ng/ListenIfAdapter.java:
```diff
@@ -0,0 +1,263 @@
+ *
+ * Author: Christian Beier <info@christianbeier.net
+ *
+ * Copyright (C) 2020 Kitchen Armor.
```
Ehr... right, so what do I put here? Is your name fine? It is YOUR software after all.

elluisian commented:
Waiting for review @bk138, of course, when you have time.

bk138 commented:
@elluisian I'll try to take a look ASAP - will take some time due to the PR's size.

freebrowser1 commented:
So, do I understand this is implemented ?
The user interface does not show any option to select network interface and the possibility to use localhost only, so this VNC is only accessible via an SSH tunnel, which is more secure.
Opening droidVNC in a public wifi is insecure as it is sniffable.

bk138 commented:
@freebrowser1 No, it's not. I didn't find time for a thorough review yet.

bk138 commented:
@elluisian thanks for the offer - yes you can rebase, that would help. I have not forgotten about this PR, otherwise I would have closed it. It's simply a matter of urgencies/priorities :-)

elluisian commented:
Hi @bk138 , I've improved the pull request, from 11 commits to just 7, moreover, they are a bit more "organized".
In particular, the first 5 commits can be stored in any order.

elluisian commented:
Hi @bk138, after a long time, I updated my fork to the last changes.
How about applying the pull request for the end of the year?
Don't worry, I'm just kidding. :P

In all seriousness, the current fork now integrates all your changes since the last time.
If there is still something to do, let me know and I will look into it.


Happy new year to you and others.
